backend/routeplanner/planner/domain/a_star.py

```diff
@@ -1,9 +1,142 @@
 import osmnx as ox
 import networkx as nx
-from .map import get_graph
+from . import router
+from .map import get_graph, load_graph
+from .priority_queue import PriorityQueue
+
+from .router import distance
+
+
+class AStarNode:
+
+    def __init__(self, node, previous, cost, id):
+        self.node = node
+        self.id = id
+        self.previous = previous
+        self.start_dist = 0  # Distance to start node
+        self.end_dist = 0  # Distance to goal node
+        self.time = 0
+        self.left_cost = 0  # Left turns cost
+        self.cost = cost  # Total cost
+
+
+    def __eq__(self, other):
+        return self.id == other.id
+
+
+    def __lt__(self, other):
+        return self.id < other.id
+
+
+    def __le__(self, other):
+        return self.id <= other.id
+
+
+    def __gt__(self, other):
+        return self.id > other.id
+
+
+
 
 def find_path(node1, node2):
-    # TODO: Implement a star algorithm here, function should return ordered list of nodes in path
-    # The path should not include first node
-    # graph = get_graph()
-    return []
\ No newline at end of file
+
+    graph= load_graph()
+
+    start_id = ox.get_nearest_node(graph, node1)
+    start_node = graph.nodes[start_id]
+
+    end_id = ox.get_nearest_node(graph, node2)
+    end_node = graph.nodes[end_id]
+
+    open = PriorityQueue()
+    closed = {}
+
+    start = AStarNode(start_node, None, 0, start_id)
+    end = AStarNode(end_node, None, 0, end_id)
+    open.put(start, 0)
+    result = []
+    while not open.empty():
+
+        #print('Open count: %d, Closed count: %d' % (len(open.elements), len(closed)))
+        current_node = open.get()
+        closed[current_node.id] = current_node
+
+        if (current_node.id == end.id):
+            path = []
+            while current_node.id != start.id:
+                # y to lat a y to lng
+                #print((current_node.node['y'], current_node.node['x']))
+                path.append((current_node.node['y'], current_node.node['x']))
+                current_node = current_node.previous
+            result = path[::-1]
+            return result
+
+        neighbors = find_neighbors(current_node, graph)
+        for next_id in neighbors:
+            neighbor = AStarNode(graph.nodes[next_id], current_node, 0, next_id)
+            neighbor.time = current_node.time + graph.edges[(current_node.id, neighbor.id, 0)]['travel_time']
+
+            if (current_node.previous is not None):
+                neighbor.left_cost = current_node.left_cost + check_turn(current_node, neighbor)
+            else:
+                neighbor.left_cost = current_node.left_cost + check_first_turn(current_node.node, neighbor.node)
```

Comment:
Chyba tu trzeba dać po prostu 0 jednak, bo to że punkt jest z lewej to się nie liczy jako kosztowny skręt w lewo?

Reply:
Tak, zgadzam się 👌